blog: Add Kubernetes-in-Kubernetes and PXE bootable servers farm blog post (#29676)
kubernetes/website: content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

---
layout: blog
title: "Kubernetes-in-Kubernetes and PXE bootable servers farm"
date: 2021-09-10T22:47:55+02:00
---

**Author**: Andrei Kvapil (WEDOS)

![Kubernetes in Kubernetes](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/godeeper.png)

> When you have two own data centers, thousands of physical servers, virtual machines and hosting for hundreds of thousands sites, Kubernetes can actually simplify the management of all these things. As practice has shown, by using Kubernetes, you can declaratively describe and manage not only applications, but also the infrastructure itself. I work for the largest Czech hosting provider **WEDOS Internet a.s** and today I'll show you two of my projects — [Kubernetes-in-Kubernetes](https://github.com/kvaps/kubernetes-in-kubernetes) and [Kubefarm](https://github.com/kvaps/kubefarm).

Reviewer: ```suggestion
When you own two data centers, thousands of physical servers, virtual machines and hosting for hundreds of thousands sites, Kubernetes can actually simplify the management of all these things. As practice has shown, by using Kubernetes, you can declaratively describe and manage not only applications, but also the infrastructure itself. I work for the largest Czech hosting provider **WEDOS Internet a.s** and today I'll show you two of my projects — [Kubernetes-in-Kubernetes](https://github.com/kvaps/kubernetes-in-kubernetes) and [Kubefarm](https://github.com/kvaps/kubefarm).
```
A minor change that you might want to add. Just a grammar enhancement in the sentence :)

> With their help you can deploy a fully working Kubernetes cluster inside another Kubernetes using Helm just in a couple of commands. How and why?

Reviewer: ```suggestion
With their help you can deploy a fully working Kubernetes cluster inside another Kubernetes using Helm in just a couple of commands. How and why?
```

> Let me introduce you how our infrastructure works. All our physical servers can be divided into two groups: **control-plane** and **compute** nodes. Control plane nodes are usually set up manually, have a stable OS installed, and designed to run all cluster services, including Kubernetes control-plane. The main task of these nodes is to ensure the smooth operation of the cluster itself. Compute nodes do not have any operating system installed by default, Instead they are booting the OS image over the network directly from the control plane nodes. Their work is to carry out the workload.

Reviewer: ```suggestion
Let me introduce you to how our infrastructure works. All our physical servers can be divided into two groups : **control-plane** and **compute** nodes. Control plane nodes are usually set up manually, have a stable OS installed, and designed to run all cluster services including Kubernetes control-plane. The main task of these nodes is to ensure the smooth operation of the cluster itself. Compute nodes do not have any operating system installed by default, instead they are booting the OS image over the network directly from the control plane nodes. Their work is to carry out the workload.
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme02.jpg)

After booting, the first thing our nodes do is join to the existing Kubernetes cluster, namely, execute the **kubeadm join** command. So, kube-scheduler could schedule some pods on them and launch various workloads afterwards. From the begining we used the scheme when nodes were joined into the same cluster used for the control-plane nodes.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme03.jpg)

This scheme worked stably for over two years. However later we decided to add containerized Kubernetes to it. And now we can spawn new Kubernetes-clusters very easely right on our control-plane nodes which are now is members special admin-cluster. And compute nodes now can be joined directly to their own clusters - depending on the configuration.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme04.jpg)

## Kubefarm

This project came about with the goal to make anyone could deploy such an infrastructure just in a couple of commands using Helm and get about the same in the end.

> At this time, we moved away from the idea of a monocluster. Because it turned out to be not very convenient for managing several development teams work in the same cluster. The fact is that Kubernetes was never designed as a multi-tenant solution and at the moment it does not provide sufficient means of isolation between projects. Therefore, running separate clusters for each team turned out to be a good idea. However, there should not be too many clusters to let them be convenient to manage. Nor is it too small to have sufficient independence between development teams.

Reviewer: ```suggestion
At this time, we moved away from the idea of a monocluster. Because it turned out to be not very convenient for managing work of several development teams in the same cluster. The fact is that Kubernetes was never designed as a multi-tenant solution and at the moment it does not provide sufficient means of isolation between projects. Therefore, running separate clusters for each team turned out to be a good idea. However, there should not be too many clusters, to let them be convenient to manage. Nor is it too small to have sufficient independence between development teams.
```

> The scalability of our clusters became noticeably better after that change. The more clusters you have per number of nodes, the smaller the failure domain and the more stable they work. And as a bonus, we got fully declaratively described infrastructure. Thus, now you can deploy a new Kubernetes cluster in the same way deploying any other application in Kubernetes.

Reviewer: ```suggestion
The scalability of our clusters became noticeably better after that change. The more clusters you have per number of nodes, the smaller the failure domain and the more stable they work. And as a bonus, we got a fully declaratively described infrastructure. Thus, now you can deploy a new Kubernetes cluster in the same way as deploying any other application in Kubernetes.
```

It uses [Kubernetes-in-Kubernetes](http://github.com/kvaps/kubernetes-in-kubernetes) as a basis, [LTSP](https://github.com/ltsp/ltsp/) as PXE-server from which the nodes are booted, and automates the DHCP server configuration using [dnsmasq-controller](https://github.com/kvaps/dnsmasq-controller):

![Kubefarm](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/kubefarm.png)

## How it works

Now let's see how it works. In general, if you look at Kubernetes as from an application perspective, you can note that it follows all the principles of [The Twelve-Factor App](https://12factor.net/), and is actually written very well. Thus, it means running Kubernetes as an app in a different Kubernetes shouldn't be a big deal.

### Running Kubernetes in Kubernetes

Here we take a look at the [Kubernetes-in-Kubernetes](https://github.com/kvaps/kubernetes-in-kubernetes) project, which provides a ready-made Helm chart for running Kubernetes in Kubernetes. 

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/github.jpg)

Let's take a look at the parameters that we pass to Helm in the values file:

> * [**kubernetes/values.yaml**](https://github.com/kvaps/kubernetes-in-kubernetes/blob/master/deploy/helm/kubernetes/values.yaml)

Reviewer 2: For this link, and maybe a couple of others, I'd personally link to a release tag or some other ref that isn't _master_. (Is it obvious why? If not, I can explain).

<img align="right" alt="Kubernetes is just a five binaries" style="height: 280px;" src="/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/5binaries.png">

> Beside **persistence** (storage parameters for the cluster), the Kubernetes control-plane components are described here: namely: **etcd-cluster**, **apiserver**, **controller-manager** and **scheduler**. These are pretty much standard Kubernetes components. There is a joke that «Kubernetes is just a five binaries». So here the configuration for these binaries is located.

Reviewer 2: ```suggestion
Beside **persistence** (storage parameters for the cluster), the Kubernetes control-plane components are described here: namely: **etcd cluster**, **apiserver**, **controller-manager** and **scheduler**. These are pretty much standard Kubernetes components. There is a light-hearted saying that “Kubernetes is just five binaries”. So here is where the configuration for these binaries is located.
```

If you ever tried to bootstrap a cluster using kubeadm, then this config will remind you it's configuration. But in addition to Kubernetes entities, we also have an admin container. In fact, it is a container, which holds two binaries inside: **kubectl** and **kubeadm**. They are used to generate kubeconfig for the above components and to perform the initial configuration for the cluster. Also, in emergency case, you can always exec into it to check and manage your cluster.

After the release [has been deployed](https://asciinema.org/a/407280), we will see a list of pods: **admin-container**, **apiserver** in two replicas, **controller-manager**, **etcd-cluster**, **scheduller** and the initial job that initializes our cluster. In the end we have a command, which allows you to get shell into the admin container, we can use it to see what is happening inside:

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot01.jpg)

> Also, let's take look at the certificates. If you've ever installed Kubernetes, then you know that it has a <irony>*scary* directory</irony> **/etc/kubernetes/pki** with a bunch of some certificates. In our case, we have fully automated management of them with cert-manager. Thus it is enough to pass all certificates paramaters to Helm during installation, and all the certificates will automatically be generated for our cluster.

Reviewer: ```suggestion
Also, let's take look at the certificates. If you've ever installed Kubernetes, then you know that it has a <irony>*scary* directory</irony> **/etc/kubernetes/pki** with a bunch of some certificates. In our case, we have fully automated management of them with cert-manager. Thus, it is enough to pass all certificates parameters to Helm during installation, and all the certificates will automatically be generated for our cluster.
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot02.jpg)

> Looking at one of the certificates eg. apiserver, you can see that it has a list of DNS names and IP addresses. If you want to make this cluster accessible outside, then just describe the additional DNS names in the values file and update the release. This will update the certificate resource, and cert-manager will regeneraate the certificate. You'll no longer need to think about this. If kubeadm certificates need to be renewed at least once a year, here the cert-manager will take care and automatically renew them.

Reviewer: ```suggestion
Looking at one of the certificates, eg. apiserver, you can see that it has a list of DNS names and IP addresses. If you want to make this cluster accessible outside, then just describe the additional DNS names in the values file and update the release. This will update the certificate resource, and cert-manager will regenerate the certificate. You'll no longer need to think about this. If kubeadm certificates need to be renewed at least once a year, here the cert-manager will take care and automatically renew them.
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot04.jpg)

Here is the [diagram of the deployed cluster](https://kvaps.github.io/images/posts/Kubernetes-in-Kubernetes-%D0%B8-%D1%84%D0%B5%D1%80%D0%BC%D0%B0-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80%D0%BE%D0%B2-%D1%81-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%BE%D0%B9-%D0%BF%D0%BE-PXE/Argo_CD_kink_network.html). You can see services for all Kubernetes components: **apiserver**, **controller-manager**, **etcd-cluster** and **scheduler**. And the pods on right side to which they forward traffic.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/argocd01.jpg)

*By the way, the diagram, is drawn in [ArgoCD](https://argoproj.github.io/argo-cd/) — the GitOps tool we use to manage our clusters, and cool diagrams are one of its features.*

### Orchestrating physical servers

OK, now we have a Kubernetes control-plane deployed, but what about worker nodes, how are we going to add them? As I already said, all our servers are bare metal. We do not use virtualization to run Kubernetes, but we orchestrate all physical servers by ourselves.

> Also we do use Linux network boot feature very actively. Moreover, this is exactly the booting, not some kind of automation of the installation. When the nodes are booting, they just run a ready-made system image for them. That is, to update any node, we just need to reboot it - and it will download a new image. It is very easy, simple and convenient.

Reviewer: ```suggestion
Also, we do use Linux network boot feature very actively. Moreover, this is exactly the booting, not some kind of automation of the installation. When the nodes are booting, they just run a ready-made system image for them. That is, to update any node, we just need to reboot it - and it will download a new image. It is very easy, simple and convenient.
```

### Demo

Go to [examples](https://github.com/kvaps/kubefarm/tree/master/examples) and update our previous deployed chart to Kubefarm. Just use the [generic](https://github.com/kvaps/kubefarm/tree/master/examples/generic) parameters and look at the pods. We can see that we have added a PXE server and one more job. This job essentially goes to the deployed Kubernetes cluster and creates a new token. Now it will run repeatedly every 12 hours to generate a new token, so that the nodes can connect to our cluster.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot05.jpg)

In a [graphical representation](https://kvaps.github.io/images/posts/Kubernetes-in-Kubernetes-%D0%B8-%D1%84%D0%B5%D1%80%D0%BC%D0%B0-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80%D0%BE%D0%B2-%D1%81-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%BE%D0%B9-%D0%BF%D0%BE-PXE/Argo_CD_Applications_kubefarm-network.html), it looks about the same, but now our apiserver started to be exposed outside. 

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/argocd02.jpg)

> In the diagram, the IP is highlighted in green, our PXE server can be reached through it. At the moment, Kubernetes does not allow creating a single LoadBalancer service for TCP and UDP protocols by default, so we have to create two different services with the same IP address. One is for TFTP, and the second for HTTP, through which, the system image is downloaded.

Reviewer: ```suggestion
In the diagram, the IP is highlighted in green, our PXE server can be reached through it. At the moment, Kubernetes does not allow creating a single LoadBalancer service for TCP and UDP protocols by default, so we have to create two different services with the same IP address. One is for TFTP, and the second for HTTP, through which the system image is downloaded.
```

But this simple example is not always enough, sometimes you might need to modify the logic at boot. For example, here is a directory [advanced_network](https://github.com/kvaps/kubefarm/tree/master/examples/advanced_network), inside which there is a [values file](https://github.com/kvaps/kubefarm/tree/master/examples/advanced_network) with a simple shell script. We call it network.sh:

* [**network.sh**](https://github.com/kvaps/kubefarm/blob/3bd90666cfe367e38ee16aa5d6f6f39747874a17/examples/advanced_network/values.yaml#L14-L78)

> All this script does is take environment variables at boot time, and generates a network configuration based on them. It creates a directory and puts the netplan config inside. For example, a bonding interface is created here. Basically, this script can contain everything you need. It can hold the network configuration or generate the system services, add some hooks or describe any other logic. Anything that can be described in bash or shell languages will works here, and it will be executed at boot time.

Reviewer: ```suggestion
All this script does is take environment variables at boot time, and generates a network configuration based on them. It creates a directory and puts the netplan config inside. For example, a bonding interface is created here. Basically, this script can contain everything you need. It can hold the network configuration or generate the system services, add some hooks or describe any other logic. Anything that can be described in bash or shell languages will work here, and it will be executed at boot time.
```

Let's see how it can be [deployed](https://asciinema.org/a/407284). Let's pass the generic values file as the first parameter, and an additional values file as the second parameter. This is a standard Helm feature. This way you can also pass the secrets, but in our case, the configuration is just expanded by the second file: 

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot06.jpg)

Let's look at the configmap **foo-kubernetes-ltsp** for our netboot server and make sure that our **network.sh** script is really there. These commands used to configure the network at boot time:

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot07.jpg)

[Here](https://asciinema.org/a/407286) you can see how it works in principle. The chassis interface (we use HPE Moonshots 1500) have the nodes, you can enter **show node list** command to get a list of all the nodes. Now we are going to boot them.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot08.jpg)

> You can also get their MAC addresses by **show node macaddr all** command. We have a clever operator that collects MAC-addresses from chassis automatically and passes them to the DHCP server. Actually it's just creating custom configuration resources for dnsmasq-controller which is running in same admin Kubernetes cluster. Also here we can control the nodes themselves, e.g. turn them on and off.

Reviewer: ```suggestion
You can also get their MAC addresses by **show node macaddr all** command. We have a clever operator that collects MAC-addresses from chassis automatically and passes them to the DHCP server. Actually, it's just creating custom configuration resources for dnsmasq-controller which is running in same admin Kubernetes cluster. Also, here we can control the nodes themselves, e.g. turn them on and off.
```

If you have no such opportunity to enter the chassis through iLO and collect a list of MAC addresses for your nodes, you can consider using [catchall cluster](https://asciinema.org/a/407287) pattern. Purely speaking, it is just a cluster with a dynamic DHCP pool. Thus, all nodes that are not described in the configuration to other clusters will automatically join to this cluster.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot09.jpg)

> For example here we have a catchall cluster with some nodes. They are joined to the cluster with an auto-generated name based on their MAC address. We can connect to them and see what happens there. Here you can somehow prepare them, for example, set up the file system and then rejoin them to another cluster. 

Reviewer: ```suggestion
For example, here we have a catchall cluster with some nodes. They are joined to the cluster with an auto-generated name based on their MAC address. We can connect to them and see what happens there. Here you can somehow prepare them, for example, set up the file system and then rejoin them to another cluster. 
```

> Now let's try connecting to the node terminal and see how it is booting. After the BIOS, the network card is configured, here it sends a request to the DHCP server from specific MAC address, which redirects it to a specific PXE server. Later the kernel and initrd image are downloaded from the server using the standard HTTP protocol:

Reviewer: ```suggestion
Now let's try connecting to the node terminal and see how it is booting. After the BIOS, the network card is configured, here it sends a request to the DHCP server from a specific MAC address, which redirects it to a specific PXE server. Later the kernel and initrd image are downloaded from the server using the standard HTTP protocol:
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot10.jpg)

After loading the kernel, the node downloads the rootfs image and transfers control to systemd. Then the booting proceeds as usual, and after that the node joins Kubernetes: 

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot11.jpg)

If you take a look at **fstab**, you can see only two entries there: **/var/lib/docker** and **/var/lib/kubelet**, they are mounted as **tmpfs** (in fact, from RAM). At the same time, the root partition is mounted as **overlayfs**, so all changes that you make here on the system will be lost on the next reboot. 

> Looking into the block devices on the node, you can see some nvme disk, but it has not yet been mounted anywhere. There is also a loop device - this is the exactly rootfs image downloaded from the server. At the moment it is located in RAM, occupies 653 MB and mounted with the **loop** option.

Reviewer: ```suggestion
Looking into the block devices on the node, you can see some nvme disk, but it has not yet been mounted anywhere. There is also a loop device - this is the exact rootfs image downloaded from the server. At the moment it is located in RAM, occupies 653 MB and mounted with the **loop** option.
```

If we look in **/etc/ltsp**, we find our **network.sh** file that was executed at boot. From containers, we see running kube-proxy and pause container for it.

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot12.jpg)

## Details

### Network Boot Image

> But where does the main image come from? There is a little trick here. The image for the nodes is built through the [Dockerfile](https://github.com/kvaps/kubefarm/tree/master/build/ltsp) along with the server. The [Docker multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) feature allows you to easily add any packages and kernel modules exactly at the stage of the image build. It looks like this:

Reviewer: ```suggestion
But where does the main image come from? there is a little trick here. The image for the nodes is built through the [Dockerfile](https://github.com/kvaps/kubefarm/tree/master/build/ltsp) along with the server. The [Docker multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) feature allows you to easily add any packages and kernel modules exactly at the stage of the image build. It looks like this:
```

Author: @Mr-DG-Wick thank you for review!
Could you explain why is this correct, please?

Reviewer: In general English grammar, Capitalization in the middle of a sentence is only permitted for the letter "i" that too only for certain words that form with the letter, like I, I'll, I'm etc. So I thought maybe you forgot to change that. 

Or maybe its some stylistic preference that I'm not aware of. So, you may or may not change that. Its **optional**

Author: Isn't it a new sentence?

Author: Hey. I just thinking is it common situation in English when question mark is not ends a sentence?

Reviewer: see this @kvaps https://www.thepunctuationguide.com/terminal-points.html#:~:text=Rule%3A%20The%20question%20mark%20ends%20the%20sentence%3B%20no%20additional%20period.&text=Rule%3A%20The%20question%20mark%20(inside,sentence%3B%20no%20additional%20question%20mark.

* [**Dockerfile**](https://github.com/kvaps/kubefarm/blob/master/build/ltsp/Dockerfile)

> What's going on here? First, we take a regular Ubuntu 20.04 and install all the packages we need. First of all we install the **kernel**, **lvm**, **systemd**, **ssh**. In general, everything that you want to see on the final node should be described here. Here we also install **docker** with **kubelet** and **kubeadm**, which are used to join the node to the cluster.

Reviewer: ```suggestion
What's going on here? first, we take a regular Ubuntu 20.04 and install all the packages we need. First of all we install the **kernel**, **lvm**, **systemd**, **ssh**. In general, everything that you want to see on the final node should be described here. Here we also install **docker** with **kubelet** and **kubeadm**, which are used to join the node to the cluster.
```

Author: Same here

> And then we make an additional configuration. In the last stage, we simply install **tftp** with **nginx** (which serves our image to clients), **grub** (bootloader). Then root of the previous stages copied into the final image and generate squashed image from it. That is, in fact, we get a docker image, which has both the server and the boot image for our nodes. At the same time, by changing the Dockerfile, it can be easily updated.

Reviewer: ```suggestion
And then we make an additional configuration. In the last stage, we simply install **tftp** with **nginx** (which serves our image to clients), **grub** (bootloader). Then root of the previous stages copied into the final image and generate squashed image from it. That is, in fact, we get a docker image, which has both the server and the boot image for our nodes. At the same time, it can be easily updated by changing the Dockerfile.
```

### Webhooks and API aggregation layer

I want to pay special attention to the problem of webhooks and aggregation layer. In general, webhooks is a Kubernetes feature that allows you to respond to the creation or modification of any resources. Thus, you can add a handler so that when resources are applied, Kubernetes must send request to some pod and check if configuration of this resource is correct, or make additional changes to it.

> But the point is, in order for the webhooks to work, the apiserver must have direct access to the cluster for which it is running. And if it is started in a separate cluster, like our case, or even separately from the any cluster, then Konnectivity service can help us here. Konnectivity is one of the optional but officially supported Kubernetes components.

Reviewer: `the`?
You wanted to mention a special cluster there?. If so, then you may want to change that here.

> Let's take cluster of four nodes for example, each of them is running a **kubelet** and we have other Kubernetes components running outside: **apiserver**, **scheduller** and **controller-manager**. By default, all these components interact with the apiserver directly - this is the most known part of the Kubernetes logic. But in fact, there is also a reverse connections. For example, when you want to view the logs or run a **kubectl exec** command, apiserver establishes a connection to the specific kubelet independently:

Reviewer: ```suggestion
Let's take cluster of four nodes for example, each of them is running a **kubelet** and we have other Kubernetes components running outside: **apiserver**, **scheduller** and **controller-manager**. By default, all these components interact with the apiserver directly - this is the most known part of the Kubernetes logic. But in fact, there is also a reverse connection. For example, when you want to view the logs or run a **kubectl exec** command, apiserver establishes a connection to the specific kubelet independently:
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/konnectivity01.jpg)

> But the problem is that if we have a webhook, then it is usually runs as a standard pod with a service in our cluster. And when apiserver tries to reach it, it will fail because it will try to access an in-cluster service named **webhook.namespace.svc** being outside of the cluster where it is actually running:

Reviewer: ```suggestion
But the problem is that if we have a webhook, then it usually runs as a standard pod with a service in our cluster. And when apiserver tries to reach it, it will fail because it will try to access an in-cluster service named **webhook.namespace.svc** being outside of the cluster where it is actually running:
```

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/konnectivity02.jpg)

And here Konnectivity comes to our rescue. Konnectivity is a tricky proxy server developed especially for Kubernetes. It can be deployed as a server next to the apiserver. And Konnectivity-agent is deployed in several replicas directly in the cluster you want to access. The agent establishes a connection to the server and set up a stable channel to make apiserver able to access all webhooks and all kubelets in the cluster. Thus, now all communication with the cluster will take place through the Konnectivity-server:

![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/konnectivity03.jpg)

## Our plans

> Of course, we are not going to stop at this stage. People interested in the project often write to me. And if there will be a sufficient number of interested people, I hope to move Kubernetes-in-Kubernetes project under [Kubernetes SIGs](https://github.com/kubernetes-sigs), by representing it in form of the official Kubernetes Helm chart. Perhaps, by making this project independed we'll gather an even larger community.

Reviewer: ```suggestion
Of course, we are not going to stop at this stage. People interested in the project often write to me. And if there will be a sufficient number of interested people, I hope to move Kubernetes-in-Kubernetes project under [Kubernetes SIGs](https://github.com/kubernetes-sigs), by representing it in form of the official Kubernetes Helm chart. Perhaps, by making this project independent we'll gather an even larger community.
```

I am also thinking of integrating it with the Machine Controller Manager, which would allow creating worker nodes, not only of physical servers, but for example creating virtual machines using kubevirt and running them in the same Kubernetes cluster. By the way, it also allow to spwan virtual machines in the clouds, and having a control-plane deployed locally.

> I am also considering the option of integrating with the Cluster-API so that you can create physical Kubefarm clusters directly through the Kubernetes environment. But at the moment I'm not completely sure about this idea. If you have any thoughts on this matter, I will be happy to listen them.

Reviewer: ```suggestion
I am also considering the option of integrating with the Cluster-API so that you can create physical Kubefarm clusters directly through the Kubernetes environment. But at the moment I'm not completely sure about this idea. If you have any thoughts on this matter, I'll be happy to listen to them.
```
